### DISCUSSION TITLE
Nextflow Error

### DISCUSSION THREAD

**Discussion by gregdenay** [bio-raum/FooDMe2#22] [General]
@marchoeppner 

so I'm keeping on working on the reporting, I added a QC report for Vsearch (not in MultiQC yet), which works.
Then I did the same for DADA and I get following error:

```
Missing process or function mergers([DataflowBroadcast around DataflowStream[?]])

 -- Check script './workflows/./../subworkflows/illumina_workflow/main.nf' at line: 86 or see '.nextflow.log' file for more details
```

Do you have an idea would could be wrong? I did not actually modify the `subworkflows/illumina_workflow/main.nf` file...

I pushed the changes to the `reporting` branch to not break everything.  
New/modified files are:
`modules/helper/dada_stats`
`subworkflow/dada2`

**Comment by marchoeppner** [comment]
Ah, yes, Nextflow is notoriously terrible at alerting you to very obvious problems ;)

https://github.com/bio-raum/FooDMe2/blob/00e76626aca4856936e071f3166879a90341efb0/subworkflows/dada2/main.nf#L48

-> missing a comma at the end of the line. 

The Dada Stats module still fails, but I think  that is an issue with the Rscript. I am pushing the missing comma anyway.

**Comment by marchoeppner** [comment]
Ah, another thing that needs changing... which goes a bit deeper into Nextflow'isms:

If you want to pass two separate files into a process that need to be from the same sample, you first need to join the channels - else Nextflow will randomly combine any two files (here: chimera and mergers). 

```NEXTFLOW
ch_denoising_with_chimera = DADA2_DENOISING.out.mergers.join(DADA2_RMCHIMERA.out.rds)

// Denoising stats
HELPER_DADA_STATS(
    ch_denoising_with_chimera
)
``` 

Likewise, the module needs to be updated to only use a single input argument:

```NEXTFLOW
tuple val(meta), path(mergers), path(seqtab)  // mergers rds and seqtab
```

The join operation, by default, combines any two channels in which the first element is identical. 

For example:

``` bash
ch_a: [ [ greg, snakemake] ]
ch_b: [ [ greg, nextflow ]]

ch_join = ch_a.join(ch_b)

-> [ [ greg, snakemake, nextflow ]]
```

In our case, we use the "meta" dictionary, which happens to be the same for both channels since it never changed throughout the subworkflow. Another option would be to have a value that corresponds to a sample id as the first element. But the dictionary concept is a bit more powerful since we could update/change it throughout.

**Reply by gregdenay** [reply]
That makes sense... It's a lot to take in.

How would you go about aggregating a process output accross samples? i.e. create a channel:

```
[ [ [meta1, file1], [meta2, file2], [meta3, file3] ]
```

Is it simply `collect` like so ? 

```
ch_aggregate = PROCESS.out.json.collect()
```

**Reply by marchoeppner** [reply]
almost - you probably want to get rid of the meta objects first (since they are not needed, and depending on your downstream process might interfere with your input channel structure):

```NEXTFLOW
ch_with_metas.map { meta, json ->
    json
}.set { ch_without_meta }

SOME_PROCESS(
   ch_without_metas.collect()
)

```
and inside that process:

```NEXTFLOW
[...]
input:
path(jsons)

output:
[...]

```

Now, if you were to use a script that finds the jsons by itself (like pythons glob.glob("*.json")) - it wouldn't be critical to do it this way of course. Otherwise your ` path(jsons)`  object would be a wild mix of jsons and meta objects and would break any script that just wants "jsons" as input argument. If that makes sense? ;)

**Reply by gregdenay** [reply]
It makes sense, the channel approach seems cleaner than a glob script though. 
If you remove the meta object, you'd have to parse the filenames to retrieve the sample id, right? Is that the Nextflow approach?

**Reply by marchoeppner** [reply]
Well, there is no strict nextflow approach, I suppose. For Jsons, one could encode the sample name in the JSON object to track sample names. Else, yes, one could use the file name - I do that too on occasion, although it feels a little "dodgy".

**Reply by gregdenay** [reply]
I somehow don't manage to use the collected channel in a Python script. 

I have in the nextflow workflow:
```
    /*
    Clustering statistics (per-sampel)
    */
    HELPER_VSEARCH_STATS(
        ch_reporting
    )

    HELPER_VSEARCH_STATS.out.json.map { meta, json ->
        json
    }.set { ch_json_nometa }
    ch_json_nometa.view()

    /*
    MultiQC report
    */
    HELPER_VSEARCH_MULTIQC(
        ch_json_nometa.collect()
    )
```

the `HELPER_VSEARCH_MULTIQC` looks like this:

```
process HELPER_VSEARCH_MULTIQC {
    input:
    path(jsons) // json reports

    output:
    path('vsearch_mqc.json'), emit: json

    script:

    """
    #!/usr/bin/env python

    import json

    data = {}
    for j in $jsons:
        with open(j, "r") as fi:
            data.update(json.load(fi))
   # do more stuff
```

but the path list form Groovy ist formatted as `filename1 filename2 filename3`, which is not a valid python list.
According to the Nextflow doc I would have thought that the format was correct, did I miss something here?

**Reply by marchoeppner** [reply]
But the information emitted from:

```
    ch_json_nometa.view()
```

looks as expected? Can you check if the files are correctly staged in the process directory?

The `filename` placeholder typically means that it didn't actually see files in the input.

**Reply by gregdenay** [reply]
The `view()` output is :
```
/home/debian/NGS/spezies_indev/FooDMe2/work/cc/37851b20adee211aa068a4161bfdff/ERR10436150.vsearch_stats.json
/home/debian/NGS/spezies_indev/FooDMe2/work/f2/1ce30ede5443f2d6e3c3d75ce87338/ERR10436141.vsearch_stats.json
/home/debian/NGS/spezies_indev/FooDMe2/work/69/0a2bdde0b4da53516f333af93cf255/ERR10436143.vsearch_stats.json
/home/debian/NGS/spezies_indev/FooDMe2/work/9c/04caa3d190f3f6d1b7c941fc216b7d/ERR10436139.vsearch_stats.json
```
-> They seem to come from different folders ? is that expected?

The Python Error is:

```
Command error:
    File ".command.sh", line 6
      for j in ERR10436150.vsearch_stats.json ERR10436141.vsearch_stats.json ERR10436143.vsearch_stats.json ERR10436139.vsearch_stats.json:
                                              ^^^^^^^^^^^
  SyntaxError: invalid syntax
```

All inputs are staged correctly

**Reply by marchoeppner** [reply]
The folders being different is expected, Nextflow sym links files from one process to the next. 

So this is probably a problem with the python code trying to read the Nextflow list. Basically, it is not a list/array but gets simply converted into a string like "fileA fileB fileC". I am guessing python expects a list here.

**Reply by marchoeppner** [reply]
I suppose you could turn it into a python  list?

(code not checked)

list = $json.split(" ")

**Reply by gregdenay** [reply]
fells like a hack but it works. I got hung up on thinking the problem was on the nextflow side

**Reply by marchoeppner** [reply]
very much a hack - but I suppose unavoidable, since you are crossing "language barriers". Probably the one advantage that I am willing to give to Snakemake , i.e. the pipeline engine being written in the same language as the processes/rules.

**Comment by gregdenay** [comment]
I'm super slow but VSearch/Dada are now in the multiqc report :)
